webapp/channels/src/packages/mattermost-redux/src/actions/users.ts

```diff
@@ -152,49 +159,61 @@ export function getProfiles(page = 0, perPage: number = General.PROFILE_CHUNK_SI
     };
 }
 
-export function getMissingProfilesByIds(userIds: string[]): ActionFuncAsync<UserProfile[]> {
-    return async (dispatch, getState) => {
+export function getMissingProfilesByIds(userIds: string[]): ActionFuncAsync<Array<UserProfile['id']>> {
+    return async (dispatch, getState, {loaders}: any) => {
+        if (!loaders.missingStatusLoader) {
+            loaders.missingStatusLoader = new DelayedDataLoader<UserProfile['id']>({
+                fetchBatch: (userIds) => dispatch(getStatusesByIds(userIds)),
+                maxBatchSize: maxUserIdsPerProfilesRequest,
+                wait: missingProfilesWait,
+            });
+        }
+
+        if (!loaders.missingProfileLoader) {
+            loaders.missingProfileLoader = new DelayedDataLoader<UserProfile['id']>({
+                fetchBatch: (userIds) => dispatch(getProfilesByIds(userIds)),
+                maxBatchSize: maxUserIdsPerProfilesRequest,
+                wait: missingProfilesWait,
+            });
+        }
+
         const state = getState();
-        const {profiles} = state.entities.users;
-        const enabledUserStatuses = getIsUserStatusesConfigEnabled(state);
-        const missingIds: string[] = [];
-        userIds.forEach((id) => {
-            if (!profiles[id]) {
-                missingIds.push(id);
-            }
-        });
+
+        const missingIds = userIds.filter((id) => !selectUser(state, id));
 
         if (missingIds.length > 0) {
-            if (enabledUserStatuses) {
-                dispatch(getStatusesByIds(missingIds));
+            if (getIsUserStatusesConfigEnabled(state)) {
+                loaders.missingStatusLoader.queueForLoading(missingIds);
             }
-            return dispatch(getProfilesByIds(missingIds));
+
+            await loaders.missingProfileLoader.queueAndWait(missingIds);
         }
 
-        return {data: []};
+        return {
+            data: missingIds,
```

Comment:
Why do the callers need to know about the `missingIds`?

Reply:
There's one place in the "user is typing" message component (`MsgTyping`) looks at the result to determine if a user was loaded so that it knows whether or not we've already updated their status. It really only cares about the number of items loaded, so returning the IDs is as good as returning the objects for it